crates/ra_hir/src/expr/validation.rs

```diff
@@ -0,0 +1,67 @@
+use rustc_hash::FxHashSet;
+use crate::{
+    expr::AstPtr,
+    HirDatabase,
+    Function,
+    Name,
+    diagnostics::{DiagnosticSink, MissingFields},
+    adt::AdtDef,
+    Path
+};
+
+use ra_syntax::ast::{AstNode, NamedFieldList};
+
+use super::{Expr, StructLitField, ExprId};
+
+pub(crate) fn validate_body(func: &Function, db: &impl HirDatabase, sink: &mut DiagnosticSink) {
+    let body = func.body(db);
+    for e in body.exprs() {
+        match e {
+            (id, Expr::StructLit { path, fields, spread }) => {
+                validate_struct_literal(id, path, fields, spread, func, db, sink)
+            }
+            _ => (),
+        }
+    }
+}
+
+fn validate_struct_literal(
+    id: ExprId,
+    _path: &Option<Path>,
+    fields: &Vec<StructLitField>,
+    spread: &Option<ExprId>,
+    func: &Function,
+    db: &impl HirDatabase,
+    sink: &mut DiagnosticSink,
+) {
+    if let Some(_) = spread {
+        return;
+    }
+    let file_id = func.source(db).0;
+    let source_map = func.body_source_map(db);
+    let infer = func.infer(db);
+    let lit_fields: FxHashSet<_> = fields.into_iter().map(|f| &f.name).collect();
+    let struct_ty = &infer[id];
+    if let Some((AdtDef::Struct(s), _)) = struct_ty.as_adt() {
+        let missed_fields: Vec<Name> = s
+            .fields(db)
+            .into_iter()
+            .filter_map(|f| {
+                let name = f.name(db);
+                if lit_fields.contains(&name) {
+                    None
+                } else {
+                    Some(name)
+                }
+            })
+            .collect();
+        if missed_fields.is_empty() {
+            return;
+        }
+        let source_file = db.parse(file_id.original_file(db));
+        let lit_node = source_map.expr_syntax(id).unwrap().to_node(&source_file);
+        let field_list_node = lit_node.descendants().find_map(NamedFieldList::cast).unwrap();
```

Comment:
Walking through descendents is not ideal, b/c in pathalogical cases there could be quite a few of those!

What we should do here is cast lit_node to `ast::StructLit`, and use the relevant method.